syntax = "proto3";

package lca_stack;

import "google/protobuf/struct.proto";

// High-level run / agent mode for Status messages.
enum Mode {
  MODE_UNSPECIFIED = 0;
  MODE_INIT = 1;
  MODE_RUNNING = 2;
  MODE_STOPPED = 3;
}

// Event severity.
enum Severity {
  SEVERITY_UNSPECIFIED = 0;
  SEVERITY_INFO = 1;
  SEVERITY_WARNING = 2;
  SEVERITY_ERROR = 3;
}

// Common header for all envelopes. Times are generated by the originating process.
message Header {
  string run_id = 1;
  string agent_id = 2;
  uint64 seq = 3;
  uint64 t_mono_ns = 4;
  uint64 t_wall_ns = 5;
}

// Negotiation handshake carried as a standalone framed protobuf message (not an Envelope).
message Handshake {
  uint32 protocol_version = 1;
  repeated uint32 supported_schema_versions = 2;
  uint32 selected_schema_version = 3;
  string role = 4;   // daemon | adapter | autonomy
  string stage = 5;  // daemon_hello | client_hello | daemon_confirm
  string run_id = 6;
  string agent_id = 7;
  uint32 adapter_port = 8;
  uint32 autonomy_port = 9;
  string scenario = 10;
  uint32 seed = 11;
  string message = 12;
}

// Core: directed instruction to a target agent or group.
message Command {
  string command_type = 1;
  string target_agent_id = 2;
  string target_group = 3;
  google.protobuf.Struct params = 4;
}

// Core: heartbeat and health state.
message Status {
  Mode mode = 1;
  repeated string faults = 2;
  bool estop = 3;
  uint64 heartbeat_seq = 4;
  uint64 daemon_wall_ns = 5;
  uint64 last_adapter_wall_ns = 6;
  uint64 last_autonomy_wall_ns = 7;
  uint64 last_team_rx_wall_ns = 8;
  uint64 last_team_tx_wall_ns = 9;
}

// Core: coordination information (not a direct command).
message TeamMessage {
  string message_type = 1;
  google.protobuf.Struct data = 2;
}

// Core: lifecycle, safety, warnings, and other events.
message Event {
  string event_type = 1;
  Severity severity = 2;
  string message = 3;
  google.protobuf.Struct data = 4;
}

// Local adapter observation.
message LocalObservation {
  google.protobuf.Struct data = 1;
}

// Autonomy request for actuation.
message ActuationRequest {
  string target_agent_id = 1;
  uint64 expires_wall_ns = 2;
  google.protobuf.Struct data = 3;
}

// Final actuation delivered to the adapter.
message Actuation {
  bool safety_applied = 1;
  string safety_reason = 2;
  google.protobuf.Struct data = 3;
}

// Versioned message envelope.
message Envelope {
  uint32 schema_version = 1;
  Header header = 2;

  // Routing / semantics.
  string topic = 3;
  string kind = 4;

  // Cross-log identity (origin of this logical message).
  string origin_agent_id = 5;
  uint64 origin_seq = 6;
  string origin_topic = 7;

  // Normalization diagnostics.
  bool header_injected = 8;
  string header_injected_reason = 9;
  bool topic_normalized = 10;
  string topic_original = 11;

  // Optional free-form metadata.
  google.protobuf.Struct meta = 12;

  oneof payload {
    Command command = 20;
    Status status = 21;
    TeamMessage team_message = 22;
    Event event = 23;
    LocalObservation observation = 24;
    ActuationRequest actuation_request = 25;
    Actuation actuation = 26;
  }
}
